src/decode-hdlc.c

```diff
@@ -0,0 +1,122 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \ingroup decode
+ *
+ * @{
+ */
+
+
+/**
+ * \file
+ *
+ * \author Victor Julien <victor@inliniac.net>
+ *
+ * Decode Cisco HDLC
+ */
+
+#include "suricata-common.h"
+#include "decode.h"
+#include "decode-hdlc.h"
+#include "decode-events.h"
+
+#include "util-unittest.h"
+#include "util-debug.h"
+
+int DecodeHDLC(ThreadVars *tv, DecodeThreadVars *dtv, Packet *p,
+                   const uint8_t *pkt, uint32_t len)
+{
+    StatsIncr(tv, dtv->counter_hdlc);
+
+    if (unlikely(len < HDLC_HEADER_LEN)) {
+        ENGINE_SET_INVALID_EVENT(p, HDLC_PKT_TOO_SMALL);
+        return TM_ECODE_FAILED;
+    }
+
+    if (unlikely(len > HDLC_HEADER_LEN + USHRT_MAX)) {
+        return TM_ECODE_FAILED;
```

Comment:
Should an event be set in this case too?

Reply:
Good question. I don't think we do this in other places either. I wonder if we should set an event, and then limit the len we pass on to what we support (for an IP packet max it to USHRT_MAX).  The event could be something like 'unexpected trailing data'. However it would require some investigation if trailing data might be normal. For example in ethernet it can be.

Reply:
Created a ticket for this https://redmine.openinfosecfoundation.org/issues/3503